### PR TITLE
OpenAI accepts URIs as connection strings

### LINKED ISSUE #2652: OpenAI client fails to parse a connection string in client
I have a connection string as ConnectionStrings__foobar=http://localhost:11434

When in my client (using aspire-starter) and adding `builder.AddAzureOpenAI("foobar");` it fails with

```
System.ArgumentException
  HResult=0x80070057
  Message=Format of the initialization string does not conform to specification starting at index 0.
  Source=System.Data.Common
  StackTrace:
   at System.Data.Common.DbConnectionOptions.GetKeyValuePair(String connectionString, Int32 currentPosition, StringBuilder buffer, Boolean useOdbcRules, String& keyname, String& keyvalue)
   at System.Data.Common.DbConnectionOptions.ParseInternal(Dictionary`2 parsetable, String connectionString, Boolean buildChain, Dictionary`2 synonyms, Boolean firstKey)
   at System.Data.Common.DbConnectionOptions..ctor(String connectionString, Dictionary`2 synonyms, Boolean useOdbcRules)
   at System.Data.Common.DbConnectionStringBuilder.set_ConnectionString(String value)
   at Aspire.Azure.AI.OpenAI.AzureOpenAISettings.Aspire.Azure.Common.IConnectionStringSettings.ParseConnectionString(String connectionString)
   at Aspire.Azure.Common.AzureComponent`3.AddClient(IHostApplicationBuilder builder, String configurationSectionName, Action`1 configureSettings, Action`1 configureClientBuilder, String connectionName, String serviceKey)
   at Microsoft.Extensions.Hosting.AspireAzureOpenAIExtensions.AddAzureOpenAI(IHostApplicationBuilder builder, String connectionName, Action`1 configureSettings, Action`1 configureClientBuilder)
   at Program.<Main>$(String[] args) in D:\source\repos\AspireApp66\AspireApp66.Web\Program.cs:line 8
```

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #2652

Feature is already documented
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2715)

### CODE REVIEW

File: tests/Aspire.Azure.AI.OpenAI.Tests/AspireAzureAIOpenAIExtensionsTests.cs [hunk 1]
```diff
@@ -65,4 +65,32 @@ public void ConnectionStringCanBeSetInCode(bool useKeyed)
...
     }
+
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void ReadsUriFromConnectionStringsCorrectly(bool useKeyed)
+    {
+        var builder = Host.CreateEmptyApplicationBuilder(null);
+        builder.Configuration.AddInMemoryCollection([
+            new KeyValuePair<string, string?>("ConnectionStrings:openai", "https://yourservice.openai.azure.com/")
```
Comment (timheuer) [inline tests/Aspire.Azure.AI.OpenAI.Tests/AspireAzureAIOpenAIExtensionsTests.cs:76]: I would add another test with `host:port` just to be sure ;-) -- this was the URI I was using was http://foo:12345

File: src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs [hunk 2]
```diff
@@ -49,23 +49,26 @@ public sealed class AzureOpenAISettings : IConnectionStringSettings
...
-        if (connectionBuilder.ContainsKey(ConnectionStringKey))
+        else
         {
-            Key = connectionBuilder[ConnectionStringKey].ToString();
+            var connectionBuilder = new DbConnectionStringBuilder
+            {
+                ConnectionString = connectionString
+            };
+
+            if (connectionBuilder.ContainsKey(ConnectionStringEndpoint) && Uri.TryCreate(connectionBuilder[ConnectionStringEndpoint].ToString(), UriKind.Absolute, out var serviceUri))
```
Comment (timheuer) [inline src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs:63]: Wouldn't this still fail if my connectionString value was "Endpoint=http://foobar.com:12345;Key=abc123"

**Review by timheuer — COMMENTED** [review]
Made two suggestions -- I'm thinking if there happens to be a connectionstring that still is a URI with port + key that maybe this still fails at the DbConnectionBuilder?

File: src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs (hunk 2 above)
Comment (sebastienros) [reply src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs:63]: I don't think, what do you have in mind? The current tests use `"Endpoint=https://aspireopenaitests.openai.azure.com/;Key=fake"`

I added some more tests to show you.

File: tests/Aspire.Azure.AI.OpenAI.Tests/AspireAzureAIOpenAIExtensionsTests.cs (hunk 1 above)
Comment (sebastienros) [reply tests/Aspire.Azure.AI.OpenAI.Tests/AspireAzureAIOpenAIExtensionsTests.cs]: Done

**Comment by sebastienros** [comment]
Unrelated test failure

```
  Failed Aspire.Dashboard.Tests.TelemetryRepositoryTests.LogTests.Subscription_MultipleUpdates_MinExecuteIntervalApplied [499 ms]
  Error Message:
   Elapsed time 00:00:00.4949117 should be greater than min execute interval 00:00:00.5000000 on read.
Expected: True
Actual:   False
  Stack Trace:
     at Aspire.Dashboard.Tests.TelemetryRepositoryTests.LogTests.Subscription_MultipleUpdates_MinExecuteIntervalApplied() in /_/tests/Aspire.Dashboard.Tests/TelemetryRepositoryTests/LogTests.cs:line 709
--- End of stack trace from previous location ---
  Passed Aspire.Dashboard.Tests.TelemetryRepositoryTests.LogTests.AddLogs_Error_UnviewedCount_WithReadSubscriptionOneApp [< 1 ms]
  Passed Aspire.Dashboard.Tests.TelemetryRepositoryTests.LogTests.AddLogs_MultipleOutOfOrder [1 ms]
  Passed Aspire.Dashboard.Tests.TelemetryRepositoryTests.LogTests.GetLogs_UnknownApplication [< 1 ms]
```

Related: https://github.com/dotnet/aspire/pull/2703

**Review by timheuer — APPROVED** [review]
Lgtm

File: src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs [hunk 3]
```diff
@@ -49,23 +49,26 @@ public sealed class AzureOpenAISettings : IConnectionStringSettings
...
+        if (Uri.TryCreate(connectionString, UriKind.Absolute, out var uri))
         {
             Endpoint = uri;
         }
-
-        if (connectionBuilder.ContainsKey(ConnectionStringKey))
+        else
         {
-            Key = connectionBuilder[ConnectionStringKey].ToString();
+            var connectionBuilder = new DbConnectionStringBuilder
```
Comment (davidfowl) [inline src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs:58]: DbConnectionStringBuilder?

Comment (sebastienros) [reply src/Components/Aspire.Azure.AI.OpenAI/AzureOpenAISettings.cs:58]: DbConnectionStringBuilder!

**Pull request merged**